### PR TITLE
Search constraint and Install/Upgrade default constraint

### LINKED ISSUE #2270: Helm upgrade should not install pre-release version by default
If semver is followed and a user uses pre-released versions they can be installed over the stable version.

If a user has the following versions and 0.0.5 is currently installed.  A helm upgrade will install `0.1.1-alpha` when I believe most users would prefer `0.1.0` to be installed.  Possibly an additional flag to allow pre-released charts to be an upgrade path.

* 0.1.1-alpha
* 0.1.0
* 0.0.5

### PR DESCRIPTION
**Pull request by sstarcher**
Fixes #2270 

* helm search now has added support for `-v/--version` to specify a version constraint using semver
* Helm Install/Upgrade when not specifying a version will apply a semver constraint of `*` which will not return pre-release versions.  Specifying a version will allow pre-released versions to be specifically installed.

From the docs of https://github.com/masterminds/semver
```
SemVer comparisons without a pre-release value will skip pre-release versions. For example, >1.2.3 will skip pre-releases when looking at a list of values while >1.2.3-alpha.1 will evaluate pre-releases.
```

### CODE REVIEW

File: pkg/repo/index.go [hunk 1]
```diff
@@ -155,12 +155,25 @@ func (i IndexFile) Get(name, version string) (*ChartVersion, error) {
 	if len(vs) == 0 {
 		return nil, ErrNoChartVersion
 	}
+
+	var constraint *semver.Constraints
 	if len(version) == 0 {
-		return vs[0], nil
+		constraint, _ = semver.NewConstraint("*")
```
Comment (technosophos) [inline pkg/repo/index.go]: Good idea. Elegant way to enforce the SemVer stability constraint.

**Review by technosophos — APPROVED** [review]
Very nice addition! Thanks.

**Comment by technosophos** [comment]
I'm holding this PR for a day while we get the other refactors merged in. Nothing needs changing... I'm just trying to be strategic about the order in which we merge things.

**Comment by sstarcher** [comment]
@technosophos sounds good just ping me if I need to modify anything

**Comment by technosophos** [comment]
@sstarcher Can you do a quick re-run of the documentation generator and fix the conflict? Then I'll get this merged.

**Comment by sstarcher** [comment]
@technosophos done and checks passed

**Comment by technosophos** [comment]
Thanks again, @sstarcher. I'm merging it now.

**Pull request merged**